Refactor the TfJob to use K8s libraries

wackxu opened the pull request:
fix https://github.com/tensorflow/k8s/issues/206

1.refactor the `spec` package to the kubernetes standard API, all detention of our CRD is in pkg  `pkg/apis/tensorflow`  and also the validation、 set Default for the CRD and some help func.

2.for use K8s code gen, wo import the project https://github.com/kubernetes/code-generator, and some shell script in the folder `hack`. You can see how to use it in https://blog.openshift.com/kubernetes-deep-dive-code-generation-customresources/

3.All the file in `pkg/client` are all auto generated and we should not modify it manually.

4.I do not plan to change the logical to use the sharedInformer and clientset. This PR is big enough, I think I can do it in a follow PR. 

5.SharedInformer is a helpful tool with cache that we can use it to list and watch our resources. More details you can see http://borismattijssen.github.io/articles/kubernetes-informers-controllers-reflectors-stores

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/k8s/215)
<!-- Reviewable:end -->

wackxu commented:
I signed it!

jlewi commented:
/ok-to-test

jlewi commented:
Thanks for this PR. This looks like a great PR. 

The PR is very big and I'm still trying to wrap my head around it.  It would be very helpful if you could update the initial comment to provide more description about the motivation for the PR and the relevant changes. Specifically, it would be great if the initial comment in the PR could address the following questions.

1. Explain that we are using K8s code gen to generate the client.
    * Which files are auto-generated? 
2. Summarize the changes needed to support using K8s code gen
3. Explain the new package layout?
4. What is an informer and why/how you want to use them?    

---

Reviewed 6 of 35 files at r1.
Review status: 6 of 35 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[hack/update-codegen.sh, line 2 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0ANo-KDwU2U4eWwVJL:-L0ANo-KDwU2U4eWwVJM:b-ba30tb) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/hack/update-codegen.sh#L2)):*
> ```Shell
> #!/bin/bash
> ```

Please add a comment explaining what this is for?

---

*[hack/update-codegen.sh, line 28 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Atk0kMcZTopXBNIiH:-L0Atk0kMcZTopXBNIiI:b-dbal3q) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/hack/update-codegen.sh#L28)):*
> ```Shell
> #                  k8s.io/kubernetes. The output-base is needed for the generators to output into the vendor dir
> #                  instead of the $GOPATH directly. For normal projects this can be dropped.
> ${CODEGEN_PKG}/generate-groups.sh "deepcopy,client,informer,lister" \
> ```

Can these scripts for code gen be moved into their own PR to reduce the PR size?

---

*[hack/verify-codegen.sh, line 2 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0ANhDu9aj9Iwjk0Nf0:-L0ANhDu9aj9Iwjk0Nf1:b-ba30ts) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/hack/verify-codegen.sh#L2)):*
> ```Shell
> #!/bin/bash
> ```

Please add a comment explaining what this is for.

---

*[pkg/apis/tensorflow/helper/helpers.go, line 1 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0ANvgCq5HT91QL6fMt:-L0ANvgCq5HT91QL6fMu:b-87dz1t) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L1)):*
> ```Go
> package helper
> ```

Is there a package level comment somewhere?

---

*[pkg/apis/tensorflow/helper/helpers.go, line 12 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AOBGib7z8N1ki3Swm:-L0AOBGib7z8N1ki3Swn:b7drvs2) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L12)):*
> ```Go
> )
> 
> func AsOwner(tfJob *tfv1.TfJob) metav1.OwnerReference {
> ```

Exported functions should have comments.

---

*[pkg/apis/tensorflow/helper/helpers.go, line 14 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AO2wJ6tYwspMMNabJ:-L0AO2wJ6tYwspMMNabK:b-v0zg3d) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L14)):*
> ```Go
> func AsOwner(tfJob *tfv1.TfJob) metav1.OwnerReference {
> 	trueVar := true
> 	// TODO: In 1.6 this is gonna be "k8s.io/kubernetes/pkg/apis/meta/v1"
> ```

I think we can delete this TODO since we require 1.8 for CRD.

---

*[pkg/apis/tensorflow/helper/helpers.go, line 83 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Atx7HDzJ4cd_psqLb:-L0Atx7HDzJ4cd_psqLc:b-bnw1m3) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L83)):*
> ```Go
> }
> 
> // Cleanup cleans up user passed spec, e.g. defaulting, transforming fields.
> ```

Is the convention to use "Cleanup" to denote this functionality?
All TODO's should have a name attached to them. The name should be the person with the most context about the requested change. Usually this will be the author of the TODO.

---

*[pkg/apis/tensorflow/helper/helpers.go, line 95 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AOX5FVwA6iY8CfGPI:-L0AOX5GHePxyDBRvW6f:bnysqmw) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L95)):*
> ```Go
> 
> // TODO(jlewi): Get rid of the append methods that we don't need
> func AppendScalingDownCondition(cs *tfv1.TfJobStatus, from, to int) {
> ```

I think we can delete this since we don't have a scaling down condition.

---

*[pkg/apis/tensorflow/helper/helpers.go, line 104 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AOasJWlVb2LxKQSJD:-L0AOasJWlVb2LxKQSJE:b-q5mdnm) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L104)):*
> ```Go
> }
> 
> func AppendRecoveringCondition(cs *tfv1.TfJobStatus) {
> ```

What is a recovering condition? Can you add a comment since its an exported method?

---

*[pkg/apis/tensorflow/helper/helpers.go, line 112 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AuJNedXaS2NxfwFpn:-L0AuJNedXaS2NxfwFpo:b34sf15) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L112)):*
> ```Go
> }
> 
> func AppendUpgradingCondition(cs *tfv1.TfJobStatus, to string, member string) {
> ```

Here and below, please add comments explaining what these conditions mean in terms of the TfJob semantics.

---

*[pkg/apis/tensorflow/helper/helpers.go, line 123 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AuT7bmeBPkRySXHyF:-L0AuT7bmeBPkRySXHyG:brez92i) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L123)):*
> ```Go
> }
> 
> func AppendRemovingDeadMember(cs *tfv1.TfJobStatus, name string) {
> ```

What is this condition in the context of TfJob?

---

*[pkg/apis/tensorflow/helper/helpers.go, line 159 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AuZL7JjeHh5Jcaa7G:-L0AuZL7JjeHh5Jcaa7H:bcxasa4) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L159)):*
> ```Go
> }
> 
> func scalingReason(from, to int) string {
> ```

Do we need this? TfJob currently doesn't support scaling.

---

*[pkg/apis/tensorflow/v1alpha1/defaults.go, line 1 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AukqEcmPd0WZOoZo2:-L0AukqEcmPd0WZOoZo3:b2p4lf6) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/defaults.go#L1)):*
> ```Go
> package v1alpha1
> ```

Package comment. 

---

*[pkg/apis/tensorflow/v1alpha1/defaults.go, line 11 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Ausbct6CXhWsl3ldj:-L0Ausbct6CXhWsl3ldk:b-658z81) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/defaults.go#L11)):*
> ```Go
> 
> // SetDefaults sets any unspecified values to defaults
> func SetDefaults_TfJobSpec(c *TfJobSpec) error {
> ```

Why did you make SetDefaults a pure function as opposed to making at a member function (I know that's not the correct terminalogy) of TfJobSpec?

---

*[pkg/apis/tensorflow/v1alpha1/defaults.go, line 35 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AvL9r95GoD-ewPQzp:-L0AvL9sGmoKsqmHc1Cz:bjg8czd) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/defaults.go#L35)):*
> ```Go
> 
> 		//Set the default configuration for a PS server if the user didn't specify a PodTemplateSpec
> 		if r.Template == nil && r.TfReplicaType == PS {
> ```

The code to sSetDefaults changed in http://github.com/tensorflow/k8s/pull/204 . That code started setting the TerminationPolicy. Can you update the code please?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/tensorflow/k8s/215#-:-L0Ax2Crn-pzEs3qES86:b-dgtms8)*
<!-- Sent from Reviewable.io -->

jlewi commented:
Ignore my previous comment about splitting this PR up. I looked at it more thoroughly and it looks like the bulk of it is autogenerated code I think its fine to submit  this as one PR.

Overall the code looks good. I think my main comments are to delete some of the code that is a legacy of the etcd operator that we don't need anymore.

---

Reviewed 26 of 35 files at r1.
Review status: 32 of 35 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[pkg/apis/tensorflow/v1alpha1/defaults.go, line 1 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AukqEcmPd0WZOoZo2:-L0Bfx8xo-_YON-ZLceK:b-q1ajgl) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/defaults.go#L1)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Package comment. 
</blockquote></details>

NM. There's a comment in doc.go.

---

*[pkg/apis/tensorflow/v1alpha1/types.go, line 20 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Bgen_X8k8AomIYqYf:-L0Bgen_X8k8AomIYqYg:b-xziw9y) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/types.go#L20)):*
> ```Go
> )
> 
> // +genclient
> ```

Are these annotations that the code gen client understands?

---

*[pkg/apis/tensorflow/v1alpha1/types.go, line 33 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0BgcCYbbVTa21ckzDH:-L0BgcCYbbVTa21ckzDI:b-9idpo3) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/types.go#L33)):*
> ```Go
> }
> 
> // TODO(jlewi): Need to define the actual configuration for the TensorFlow TfJob.
> ```

I think we can delete this TODO now.

---

*[pkg/apis/tensorflow/v1alpha1/types.go, line 112 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Bh5GAeyNPa9BrtFiR:-L0Bh5GAeyNPa9BrtFiS:b-hc954y) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/types.go#L112)):*
> ```Go
> // TODO(jlewi): Need to define appropriate conditions and get rid of the ones we don't need.
> const (
> 	TfJobConditionReady              = "Ready"
> ```

Conditions were copied from the etcd cluster. I think we should delete them.

---

*[pkg/apis/tensorflow/v1alpha1/types.go, line 136 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Bhgv2GoP4CIgYnDvC:-L0Bhgv2GoP4CIgYnDvD:bb1u3eh) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/types.go#L136)):*
> ```Go
> 	// ControlPuased indicates the operator pauses the control of the cluster.
> 	// TODO(jlewi): I think we can get rid of ControlPaued.
> 	ControlPaused bool `json:"controlPaused"`
> ```

I think we can delete ControlPaused. Currently there's no notion of pausing a job. If we want to add that later we can add it back.

---

*[pkg/apis/tensorflow/v1alpha1/types.go, line 139 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0BhZEXlm2PJfl0LuAd:-L0BhZEXlm2PJfl0LuAe:b-2ezibw) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/types.go#L139)):*
> ```Go
> 
> 	// Condition keeps ten most recent cluster conditions
> 	Conditions []TfJobCondition `json:"conditions"`
> ```

I think we should delete conditions. It was inherited from etcd operator and I don't think it makes sense for TfJob.

---

*[pkg/apis/tensorflow/v1alpha1/types.go, line 171 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Bht4pjKwFwD7CY9iw:-L0Bht4pjKwFwD7CY9ix:b235w3w) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/types.go#L171)):*
> ```Go
> // +resource:path=tfjobs
> 
> // TfJobList is a list of etcd clusters.
> ```

etcd -> TfJobs

---

*[pkg/apis/tensorflow/v1alpha1/types.go, line 177 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0BhvzG8-u-5TqKS6oH:-L0BhvzG8-u-5TqKS6oI:bcxjwtv) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/types.go#L177)):*
> ```Go
> 	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
> 	Metadata metav1.ListMeta `json:"metadata,omitempty"`
> 	// Items is a list of third party objects
> ```

third party objects -> TfJobs

---

*[pkg/apis/tensorflow/validation/validation.go, line 31 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0BjtOHKXLgS4KJ3MHS:-L0BjtOHKXLgS4KJ3MHT:bg0hoc3) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/validation/validation.go#L31)):*
> ```Go
> 
> 		isValidReplicaType := false
> 		for _, t := range validReplicaTypes {
> ```

Can you add the validation that was added in https://github.com/tensorflow/k8s/pull/204 to check termination policy?

---

*[pkg/client/clientset/versioned/typed/tensorflow/v1alpha1/tfjob.go, line 17 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Bm1wyJlkJdYZlDKLy:-L0Bm1wyJlkJdYZlDKLz:b-acsie5) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/client/clientset/versioned/typed/tensorflow/v1alpha1/tfjob.go#L17)):*
> ```Go
> */
> 
> package v1alpha1
> ```

This file is autogenerated right even though it doesn't have a comment to that effect?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/tensorflow/k8s/215#-:-L0BmGBMIMUoDtEAUbWC:bc547ev)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
Reviewed 35 of 35 files at r1.
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[glide.yaml, line 19 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0ClDDPtUi8-Lwcho2y:-L0ClDDPtUi8-Lwcho2z:b-wxqbd6) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/glide.yaml#L19)):*
> github.com/docker/distribution
Do we need this dependency?

---

*[pkg/apis/tensorflow/v1alpha1/register.go, line 33 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0CjjVJvalngGRBBHIi:-L0CjjVJvalngGRBBHIj:boyviap) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/register.go#L33)):*
> ```Go
>   metav1.AddToGroupVersion(scheme, SchemeGroupVersion)
>   return nil
> }
> ```

It is better to add a new line.

---

*[pkg/apis/tensorflow/v1alpha1/types.go, line 20 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Bgen_X8k8AomIYqYf:-L0Ck-lq_JmePsE29ZDp:b-ydo0an) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/types.go#L20)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Are these annotations that the code gen client understands?
</blockquote></details>

Yeah, we have to add these annotations to make code-generator works.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/tensorflow/k8s/215)*
<!-- Sent from Reviewable.io -->

jlewi commented:
Let me know when this is ready for another look.

wackxu commented:
@jlewi Sorry for the delay, I am very busy now and today night I will fix  all the comments.

wackxu commented:
When I fix it, I will call you @jlewi

wackxu commented:
@jlewi Just update the PR comment and sorry that I will addressed the comments tomorrow, It is about eleven twenty in the evening and I need to go to home.

jlewi commented:
No rush on my end. Just want to make sure you aren't blocked waiting on me.

Thank you very much for doing this.

wackxu commented:
@jlewi Addressed the comments, Can you take a look? I will refactor the logic in the PR https://github.com/tensorflow/k8s/pull/234, I think the change is big, we should have a detail look.

jlewi commented:
Reviewed 31 of 527 files at r2.
Review status: 55 of 551 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[hack/update-codegen.sh, line 2 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0ANo-KDwU2U4eWwVJL:-L0m0bv6-Pocet5L8tji:b-gbr5fk) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/hack/update-codegen.sh#L2)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Please add a comment explaining what this is for?
</blockquote></details>

Can you add a comment please?

---

*[hack/update-codegen.sh, line 28 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Atk0kMcZTopXBNIiH:-L0m0ifmv5Y2crEfmYOT:b-errpmd) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/hack/update-codegen.sh#L28)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Can these scripts for code gen be moved into their own PR to reduce the PR size?
</blockquote></details>

As I mentioned elsewhere I think it makes sense to keep them in this PR.

---

*[pkg/apis/tensorflow/helper/helpers.go, line 83 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Atx7HDzJ4cd_psqLb:-L0m0yvp2Khfh-8kSP3r:bhz5jd4) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L83)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Is the convention to use "Cleanup" to denote this functionality?
All TODO's should have a name attached to them. The name should be the person with the most context about the requested change. Usually this will be the author of the TODO.
</blockquote></details>

I think we can take care of cleaning this up in a follow on PR since this PR didn't actually add the cleanup function.

---

*[pkg/apis/tensorflow/helper/helpers.go, line 159 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AuZL7JjeHh5Jcaa7G:-L0m1B9NyRAmLd79irs9:b32w0x) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/helper/helpers.go#L159)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Do we need this? TfJob currently doesn't support scaling.
</blockquote></details>

OK

---

*[pkg/apis/tensorflow/v1alpha1/defaults.go, line 11 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0Ausbct6CXhWsl3ldj:-L0m1IOhlJRPjNkZpUFY:bc4cys1) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/defaults.go#L11)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Why did you make SetDefaults a pure function as opposed to making at a member function (I know that's not the correct terminalogy) of TfJobSpec?
</blockquote></details>

I think you missed this question?

---

*[pkg/apis/tensorflow/v1alpha1/defaults.go, line 35 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0AvL9r95GoD-ewPQzp:-L0m1Rbb0vQhbkvTEO8g:b32wtd) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/defaults.go#L35)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

The code to sSetDefaults changed in http://github.com/tensorflow/k8s/pull/204 . That code started setting the TerminationPolicy. Can you update the code please?
</blockquote></details>

ok

---

*[pkg/apis/tensorflow/v1alpha1/register.go, line 33 at r1](https://reviewable.io:443/reviews/tensorflow/k8s/215#-L0CjjVJvalngGRBBHIi:-L0m0A-HnCl3M6jR8tYR:b-iw4kjv) ([raw file](https://github.com/tensorflow/k8s/blob/79a8ead5c8d3034659d18e2fcdedd3dc5161ec6e/pkg/apis/tensorflow/v1alpha1/register.go#L33)):*
<details><summary><i>Previously, gaocegege (Ce Gao) wrote…</i></summary><blockquote>

It is better to add a new line.
</blockquote></details>

Can you add a newline please?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/tensorflow/k8s/215)*
<!-- Sent from Reviewable.io -->

jlewi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

There were a couple unresolved comments but nothing major. All the vendor changes makes reviewing this PR a pain. So I'm going to merge and we can follow up in folllow on PRs.

---

Review status: 55 of 551 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/tensorflow/k8s/215#-:-L0m39_36sU4Hlsi2dUX:bnoed5p)*
<!-- Sent from Reviewable.io -->

jlewi commented:
Actually can you fix the travis build, it looks like there is an issue with dependencies

```
$ go test -v ./pkg/...

vendor/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go:23:2: cannot find package "github.com/hashicorp/golang-lru" in any of:

	/home/travis/gopath/src/github.com/tensorflow/k8s/vendor/github.com/hashicorp/golang-lru (vendor tree)

	/home/travis/.gimme/versions/go1.8.5.linux.amd64/src/github.com/hashicorp/golang-lru (from $GOROOT)

	/home/travis/gopath/src/github.com/hashicorp/golang-lru (from $GOPATH)
```

I suspect since we check in the vendor directory we no longer need to run glide install and that might fix the issue.

jlewi commented:
I think #236 (removing glide install from travis build) should also fix the issue.

ScorpioCPH commented on pkg/apis/tensorflow/v1alpha1/types.go:
```diff
@@ -0,0 +1,195 @@
+type AcceleratorVolume struct {
+	Name      string
+	HostPath  string
+	MountPath string
```
Just FYI, there is no need to mount GPU manually if we use nvidia device-plugin and nvidia-docker.

ScorpioCPH replied:
Some useful information here: https://github.com/kubernetes/kubernetes/issues/54011

jlewi replied:
Do all Kubernetes distributions support device-plugin? As of a couple months ago I still think we needed to use the file path on ACS and some other platforms.

ScorpioCPH replied:
Yes, I think device-plugin will be beta (enabled by default) in 1.10, and NVIDIA GPU discovery feature (`alpha.kubernetes.io/nvidia-gpu`) will be deprecated in the future.

Please read this doc for more details: https://github.com/kubernetes/website/pull/6736.

ScorpioCPH commented on pkg/apis/tensorflow/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,70 @@
+// SetDefaults sets any unspecified values to defaults
+func SetDefaults_TfJobSpec(c *TfJobSpec) error {
+	if c.TfImage == "" {
+		c.TfImage = DefaultTFImage
```
Is this `TfImage` required, why should we set a default value for this?

ScorpioCPH commented on pkg/apis/tensorflow/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,70 @@
+
+		//Set the default configuration for a PS server if the user didn't specify a PodTemplateSpec
+		if r.Template == nil && r.TfReplicaType == PS {
+			setDefaultPSPodTemplateSpec(r, c.TfImage)
```
Why should we set a default value here?

ScorpioCPH commented on pkg/apis/tensorflow/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,70 @@
+		}
+
+		if string(r.TfReplicaType) == "" {
+			r.TfReplicaType = MASTER
```
IMHO, should this be required? and `MASTER` is only one in distributed TF training.

jlewi replied:
+1

zjj2wry replied:
@wackxu this should fix

ScorpioCPH commented on pkg/apis/tensorflow/v1alpha1/types.go:
```diff
@@ -0,0 +1,195 @@
+
+type TfJobSpec struct {
+	// TODO(jlewi): Can we we get rid of this and use some value from Kubernetes or a random ide.
+	RuntimeId string
```
Can we reuse metadata.UID?

jlewi replied:
Can you open up an issue for this?

ScorpioCPH replied:
ok

ScorpioCPH replied:
FYI https://github.com/tensorflow/k8s/issues/249

ScorpioCPH commented on pkg/apis/tensorflow/v1alpha1/types.go:
```diff
@@ -0,0 +1,195 @@
+	State ReplicaState `json:"state"`
+
+	// ReplicasStates provides the number of replicas in each status.
+	ReplicasStates map[ReplicaState]int
```
Nit: why there is so many `states`, State/ReplicaState/TfReplicaStatus

jlewi replied:
Can you open up an issue for this?

ScorpioCPH replied:
Sure

ScorpioCPH replied:
FYI https://github.com/tensorflow/k8s/issues/249

jlewi commented:
/test all

jlewi commented:
@ScorpioCPH Can you open up an issue with your comments about the API issues? This PR is primarily changing the existing implementation but not the API. So it would be better to address API changes in another issue.

There are already a some API related issues tagged with the [API label](https://github.com/tensorflow/k8s/labels/API).

jlewi commented:
Any idea what's going on with the test failures?

Maybe try pulling in the latest changes from master. It looks like the travis build is still running `glide install` but that should have been fixed by #236.

@wackxu I think we can merge this as soon as the test failures are fixed. Would be great to get this merged since it touches a lot of files and will unblock other refactoring.

wackxu commented:
> Why did you make SetDefaults a pure function as opposed to making at a member function (I know that's not the correct terminalogy) of TfJobSpec?

@jlewi  This style is keep to be consistent with k8s. And I have add auto generated `defaulter` to use the `SetDefaults` function.  It is referenced by auto  generated  file`generated zz_generated.defaulter.go`

wackxu commented:
Sorry for the delay, Addressed comments. PTAL. I will do all the refactor work in follow two days since I am not busy now.

zjj2wry commented on pkg/apis/tensorflow/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,77 @@
+		}
+	}
+
+	return nil
```
should not `return nil` just `return`

zjj2wry commented on pkg/apis/tensorflow/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,77 @@
+	return nil
+}
+
+func setDefault_PSPodTemplateSpec(r *TfReplicaSpec, tfImage string) {
```
seems this func not join default generator, we can move this？

wackxu replied:
@zjj2wry The `setDefault_PSPodTemplateSpec` is called by the `SetDefaults_TfJob` func. I think we should keep it.

zjj2wry replied:
you are right, i miss this~

zjj2wry commented on pkg/apis/tensorflow/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+
+// SetDefaults sets any unspecified values to defaults
```
fix this lints

ScorpioCPH replied:
Can you attach more info about lints errors :)
It maybe helpful to fix this.

zjj2wry commented:
@wackxu maybe loses some dependency~
@jlewi @ScorpioCPH this PR in order to add k8s client package, i suggest merge this when wackxu fix `defaulter` comments. And we can fix api convention in other PR.
```
1.11s$ $GOPATH/bin/goveralls -service=travis-ci -v -package ./pkg/...
exit status 1: vendor/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go:23:2: cannot find package "github.com/hashicorp/golang-lru" in any of:
	/home/travis/gopath/src/github.com/tensorflow/k8s/vendor/github.com/hashicorp/golang-lru (vendor tree)
	/home/travis/.gimme/versions/go1.8.linux.amd64/src/github.com/hashicorp/golang-lru (from $GOROOT)
	/home/travis/gopath/src/github.com/hashicorp/golang-lru (from $GOPATH)
```

jlewi commented:
@zjj2wry Seems reasonable to me. I think we can merge this as soon as the tests are fixed and fix any remaining issues in follow up PRs.

wackxu commented:
@jlewi Addressed comments, PTLA

jlewi commented:
Reviewed 1 of 527 files at r2, 1 of 6 files at r3, 1 of 1 files at r4.
Review status: 55 of 552 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/tensorflow/k8s/215)*
<!-- Sent from Reviewable.io -->

jlewi commented:
Any idea why the travis build is still failing? Not sure why build would succeed in the presubmit but not the travis build.

```
$ $GOPATH/bin/goveralls -service=travis-ci -v -package ./pkg/...
exit status 1: vendor/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go:23:2: cannot find package "github.com/hashicorp/golang-lru" in any of:
	/home/travis/gopath/src/github.com/tensorflow/k8s/vendor/github.com/hashicorp/golang-lru (vendor tree)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/github.com/hashicorp/golang-lru (from $GOROOT)
	/home/travis/gopath/src/github.com/hashicorp/golang-lru (from $GOPATH)
```

wackxu commented:
@jlewi fix  travis build fail， PTAL

zjj2wry commented:
`travis-ci` fixed, let's merge this unblock other refactor.

jlewi commented:
Thanks for the hard work.
